deepdrivemd/data/stream/aggregator_reader.py

```diff
@@ -0,0 +1,276 @@
+import adios2
+import numpy as np
+from deepdrivemd.utils import t1Dto2D
+from pathlib import Path
+from typing import List, Tuple
+from deepdrivemd.data.stream.adios_utils import ADIOS_RW_FULL_API
+
+
+class ADIOS_READER:
+    """This class is used to read the next `N` steps from an adios stream.
+
+    Attributes
+    ----------
+    adios : adios2.adios2.ADIOS
+    io : adios2.adios2.IO
+    stream : adios2.adios2.Engine
+    """
+
+    def __init__(self, fn: str, config: Path, stream_name: str):
+        """Constructor
+
+        Parameters
+        ----------
+        fn: str
+            file name of bp file or sst socket (without sst extension)
+        config : Path
+            path to `adios.xml` file
+        stream_name : str
+            name of a stream in `adios.xml` file
+        """
+        self.adios = adios2.ADIOS(str(config), True)
+        self.io = self.adios.DeclareIO(stream_name)
+        self.stream = self.io.Open(fn, adios2.Mode.Read)
+
+    def __del__(self):
+        """Destructor: clean the adios resources"""
+        self.stream.Close()
+        self.io.RemoveAllVariables()
+        self.adios.RemoveAllIOs()
+
+    def next_all(
+        self, N: int
+    ) -> Tuple[
+        int,
+        List[np.ndarray],
+        List[np.ndarray],
+        List[np.ndarray],
+        List[np.ndarray],
+        List[np.ndarray],
+        List[np.ndarray],
+    ]:
+        """Read the next `N` steps of all variables.
+
+        Parameters
+        ----------
+        N : int
+            read that many steps
+
+        Returns
+        -------
+        Tuple[int, List[np.ndarray], List[np.ndarray], List[np.ndarray], List[np.ndarray], List[np.ndarray], List[np.ndarray]]
+             a tuple consisting of:
+             actual number of steps read,
+             a list of steps,
+             a list of md5sums,
+             a list of contact_maps,
+             a list of positions,
+             a list of velocities,
+             a list of rmsds
+        """
+        CMs = []
```

Comment:
Is there a way we can clean this up to make it more general? Right now it is hard coded to work with rmsd and contact maps. A decent idea might be to define a StreamVariable class which contains the metadata here 

```
variables = {
            "step": (np.int32, 0),
            "rmsd": (np.float32, 0),
            "contact_map": (np.uint8, 1),
            "positions": (np.float32, 1),
            "velocities": (np.float32, 1),
            "md5": (str, 2),
        }
```

as well as any post process steps such as:
```
ARW.d_contact_map = np.unpackbits(ARW.d_contact_map)
ARW.d_contact_map = t1Dto2D(ARW.d_contact_map)
```

Something like this perhaps:

```
class StreamVariable:
    def __init__(self, name, dtype, structure):
        self.name = name
        self.dtype = dtype
        self.structure = structure
        
    def read_stream(self, data):
        raise NotImplementedError()
        
     # maybe a write_stream() as well?
        
class ContactMapStreamVariable(StreamVariable):
    def __init__(self, name, dtype, structure):
        super.__init__(name, dtype, structure)
        
    def read_stream(self, data):
        data = np.unpackbits(data)
        data = t1Dto2D(data)
        return data
```

Then a new application would just need to make these data classes which could be passed as a list of variables to an ADIOS_READER class like:

```
variables = [
    ContactMapStreamVariable("contact_map", np.uint8, "array"),
    ...
]

ADIOS_READER(fn, config, stream_name, variables)
```
And then the Reader class function should loop over the dict, something like:

```

data = defaultdict(list)
for i in range(N):
    status = ARW.read_step(0)
    if not status:
        break
    for var in variables:
        data[var.name].append(var.read_stream(AWR[var.name]))
    return i, dict(data)
```

Also, ADIOS_RW_FULL_API class needs to store the data arrays in a dict, or maybe in the proposed StreamVariable class since I don't see how the ARW.d_ dynamic class property creation can work for arbitrary input data types without having to change this source code for each new variable type.

This also makes it so that we don't need the `next_cm` function which would help to make this more general (since next_cm is a special case of the proposed solution)

What do you think?

Reply:
ADIOS_READER and STREAMS can be simplified. 

But ARW should not be touched: remember it works not only with files but SST streams and all the variables should be on the same level for it to work with streams, so StreamVariable logic does not work there, I do not think dictionary is a good idea either. 

I do not see a problem with dynamic d_ variables: they support arbitrary datatypes (you can pass it as an argument to the class if necessary) but since ARW class interfaces with adios, it is enough for it to handle adios kind of datatypes that are hardcoded in a class: numpy arrays, scalars, strings. 

For non-adios data you would have to have different means in any case. 
Overgeneralizing eventually makes things more complicated and unmanageable.

Reply:
Can you take a shot at simplifying the ADIOS_READER and STREAMS class?

I think defaultdict would really help there to clean up all those lists.

Yeah, I remember the adios constraints. They are pretty tricky to get around too. I'd like to take a shot at it as well. Do you have a script you are using to test this portion of the code?

We really want to be able to stream arbitrary data transformations which could consist of subsets of the positions data, contact maps, or other features. Hopefully they should be numpy arrays for the most part. It would be really nice to have the code setup in a way that makes this simple. Right now, we would have to make new STREAMS and ADIOS_READER  classes for each different application which is a lot of boilerplate code. That is the part that I want to generalize out behind a data API interface. Does it make sense?

Reply:
To test any change to the code, I simply do 'make run1" which is a 30-minute minirun that involves all the components exactly the same way as in the production but with smaller number of nodes: 12 simulations, 1 aggregator. To clean after the run assuming that the data is in ../Outputs/3, I do make clean d=3. 
Usually the first few minutes is enough to see if things are working properly. For that I use:
make watch1 d=3
to see how aggregated file grows,
make watch2 d=3
to see how subdirectories for each simulation are created
make watch3 d=0013
or 
make watch3 d=0014
to see how the logs for ML and Outlier Search are growing.

With such a code, I do not see much point in unit testing. Only the real run would show problems with adios.



I'll see what I can do with ADIOS_READER and STREAMS.

Reply:
And if you mess with ADIOS, it is better not to interrupt it but let it run for all 30 minutes because some problems might appear much later.

Reply:
Thanks, Igor. This is really helpful.

Reply:
Done with generalization of ADIOS_READER and STREAMS.

Reply:
Something is not right with the generalization: run1 seems to be running as usual but in run2 each iteration of machine learning and outlier search takes unreasonable amount of time. Perhaps, lastN is not working properly.

Reply:
I think I found the problem. Testing...

Reply:
The fix is definitely working. SC-like run has been already running for 4 hours and everything works as expected.

Reply:
This is great news, I will review by tomorrow. Thanks, Igor.